airflow/providers/tableau/operators/tableau.py

```diff
@@ -0,0 +1,144 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Optional
+
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.providers.tableau.hooks.tableau import TableauHook
+from airflow.providers.tableau.sensors.tableau_job_status import TableauJobStatusSensor
+
+RESOURCES_METHODS = {
+    'datasources': ['delete', 'download', 'refresh'],
+    'groups': ['delete'],
+    'projects': ['delete'],
+    'schedule': ['delete'],
+    'sites': ['delete'],
+    'subscriptions': ['delete'],
+    'tasks': ['delete', 'run'],
+    'users': ['remove'],
+    'workbooks': ['delete', 'download', 'refresh'],
+}
+
+
+class TableauOperator(BaseOperator):
+    """
+    Exectues a Tableau API Resource
+
+    .. seealso:: https://tableau.github.io/server-client-python/docs/api-ref
+
+    :param resource: The name of the resource to use.
+    :type resource: str
+    :param method: The name of the resource's method to execute.
+    :type method: str
+    :param resource_id: The id of resource wich will recive the action.
+    :type resource_id: str
+    :param resource_id_by: The resource field to be used as an id.
+    :type resource_id_by: Optional[str]
+    :param site_id: The id of the site where the workbook belongs to.
+    :type site_id: Optional[str]
+    :param blocking_refresh: By default the extract refresh will be blocking means it will wait until it has finished.
+    :type blocking_refresh: bool
+    :param tableau_conn_id: The :ref:`Tableau Connection id <howto/connection:tableau>`
+        containing the credentials to authenticate to the Tableau Server.
+    :type tableau_conn_id: str
+    """
+
+    def __init__(
+        self,
+        *,
+        resource: str,
+        method: str,
+        resource_id: str,
+        resource_id_by: str = 'id',
```

Comment:
What it means when you provide both `resource_id` & `resource_id_by` (Like shown in the example dag you added)? From what I see you need only 1 of them (as if user provided `resource_id_by` then you search for the id and then execute the method ). It can be confusing when both are provided isn't it?

Reply:
The tableu client uses the tableu an UUID for operations. To be compatible with `TableauRefreshWorkbookOperator` that uses the workbook`s name to get this UUID then to execute the tasks, the resource_id_by says what resource field should be use to match with resource_id to return the UUID necessary to run the tasks.
Agree, it's confusing. Should I change it to `find` and `find_by`?
Wich `find_by` will recive the resource field name to match and `find` the value to match.

Reply:
The confusing part I was referring to isn't the parameter names.
When you provide both `resource_id` & `resource_id_by` the value in `resource_id_by` will be ignored because of the condition in `_get_resource_id_by` function. I suggest to make both `resource_id` & `resource_id_by` optional parameters and make sure the user pass exactly one of them. WDYT?

Reply:
Both are necessary.  
The method `_get_resource_id_by` always is executed.
When you pass `resource_id='foo'`  and `resource='workbooks'` It will search an workbook with ID foo. 
When you pass `resource_id='foo'`, `resource_id_by='name'`  and `resource='workbooks'` It will search an workbook with name foo.
When you pass `resource_id_by='name'`, `resource='workbooks'` and `resource_id=None` It won't work because the search happen using `getattr(resource, self.resource_id_by) == self.resource_id`

That's why I ask about change to find and find_by. 

```
    def _get_resource_id_by(self, tableau_hook: TableauHook) -> str:

        if self.resource_id_by == 'id':
            return self.resource_id

        for resource in tableau_hook.get_all(resource_name=self.resource):
            if getattr(resource, self.resource_id_by) == self.resource_id:
                resource_id = resource.id
                self.log.info('Found matching with id %s', resource_id)
                return resource_id
```

Reply:
Guess now it's clearer.
**resource**: The name of the resource to use.
**method**: The name of the resource's method to execute.
**find**: The reference of resource which will receive the action.
**match_with**: The resource field name to be matched with **find** parameter.